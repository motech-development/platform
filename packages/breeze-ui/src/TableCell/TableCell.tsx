/* eslint-disable react/jsx-props-no-spreading */
import React, { FC, memo, TdHTMLAttributes } from 'react';
import styled from 'styled-components';

const baseStyles = (noWrap: boolean, align = 'inherit') => `
  padding: 10px;
  text-align: ${align};
  vertical-align: middle;
  ${noWrap ? 'white-space: nowrap;' : ''}
`;

interface ITableCell {
  $noWrap: boolean;
}

const TableDataCell = styled.td<ITableCell>`
  ${({ align, $noWrap }) => `
    ${baseStyles($noWrap, align)}
    font-weight: 300;
  `}
`;

const TableHeadCell = styled.th<ITableCell>`
  ${({ align, $noWrap }) => `
    ${baseStyles($noWrap, align)}
    font-family: 'Cabin', sans-serif;
    font-weight: 600;
  `}
`;

export interface ITableCellProps
  extends TdHTMLAttributes<HTMLTableDataCellElement> {
  as?: 'td' | 'th';
  noWrap?: boolean;
}

const TableCell: FC<ITableCellProps> = ({
  as = 'td',
  noWrap = true,
  ...rest
}) =>
  as === 'td' ? (
    <TableDataCell $noWrap={noWrap} {...rest} />
  ) : (
    <TableHeadCell $noWrap={noWrap} {...rest} />
  );

export default memo(TableCell);
