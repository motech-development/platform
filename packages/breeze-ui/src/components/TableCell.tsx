import { FC, forwardRef, ThHTMLAttributes } from 'react';
import { Box } from 'react-polymorphic-box';
import TAlignment from '../utils/alignment';
import { classNames, textAlignmentClass } from '../utils/className';

type TTableCellAs = 'td' | 'th';

export interface ITableCellProps
  extends ThHTMLAttributes<HTMLTableDataCellElement> {
  align?: TAlignment;
  as?: TTableCellAs;
  noWrap?: boolean;
}

const TableCell: FC<ITableCellProps> = forwardRef<
  HTMLTableDataCellElement,
  ITableCellProps
>(
  (
    { align = 'left', as = 'td', className = '', noWrap = true, ...rest },
    ref,
  ) => (
    <Box
      as={as}
      className={classNames(
        'px-6 py-3',
        className,
        as === 'td' ? 'bg-white bg-opacity-70' : '',
        noWrap ? 'whitespace-nowrap' : '',
        textAlignmentClass(align),
      )}
      ref={ref}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
    />
  ),
);

export default TableCell;
