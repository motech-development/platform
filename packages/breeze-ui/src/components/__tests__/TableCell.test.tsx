import { render } from '@testing-library/react';
import TableCell from '../TableCell';

describe('TableCell', () => {
  describe('as a cell', () => {
    it('should have the word wrap style applied by default', () => {
      const { container } = render(
        <table>
          <tbody>
            <tr>
              <TableCell data-testid="cell">Test</TableCell>
            </tr>
          </tbody>
        </table>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the word wrap style applied', () => {
      const { container } = render(
        <table>
          <tbody>
            <tr>
              <TableCell noWrap data-testid="cell">
                Test
              </TableCell>
            </tr>
          </tbody>
        </table>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should not have the word wrap style applied', () => {
      const { container } = render(
        <table>
          <tbody>
            <tr>
              <TableCell noWrap={false} data-testid="cell">
                Test
              </TableCell>
            </tr>
          </tbody>
        </table>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('as a header', () => {
    it('should have the word wrap style applied by default', () => {
      const { container } = render(
        <table>
          <thead>
            <tr>
              <TableCell as="th" data-testid="cell">
                Test
              </TableCell>
            </tr>
          </thead>
        </table>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the word wrap style applied', () => {
      const { container } = render(
        <table>
          <thead>
            <tr>
              <TableCell noWrap as="th" data-testid="cell">
                Test
              </TableCell>
            </tr>
          </thead>
        </table>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should not have the word wrap style applied', () => {
      const { container } = render(
        <table>
          <thead>
            <tr>
              <TableCell noWrap={false} as="th" data-testid="cell">
                Test
              </TableCell>
            </tr>
          </thead>
        </table>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
