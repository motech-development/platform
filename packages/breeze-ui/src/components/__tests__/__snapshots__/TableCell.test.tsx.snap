// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableCell as a cell should have the word wrap style applied 1`] = `
<table>
  <tbody>
    <tr>
      <td
        class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
        data-testid="cell"
      >
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableCell as a cell should have the word wrap style applied by default 1`] = `
<table>
  <tbody>
    <tr>
      <td
        class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
        data-testid="cell"
      >
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableCell as a cell should not have the word wrap style applied 1`] = `
<table>
  <tbody>
    <tr>
      <td
        class="px-6 py-3 bg-white bg-opacity-70 text-left"
        data-testid="cell"
      >
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableCell as a header should have the word wrap style applied 1`] = `
<table>
  <thead>
    <tr>
      <th
        class="px-6 py-3 whitespace-nowrap text-left"
        data-testid="cell"
      >
        Test
      </th>
    </tr>
  </thead>
</table>
`;

exports[`TableCell as a header should have the word wrap style applied by default 1`] = `
<table>
  <thead>
    <tr>
      <th
        class="px-6 py-3 whitespace-nowrap text-left"
        data-testid="cell"
      >
        Test
      </th>
    </tr>
  </thead>
</table>
`;

exports[`TableCell as a header should not have the word wrap style applied 1`] = `
<table>
  <thead>
    <tr>
      <th
        class="px-6 py-3 text-left"
        data-testid="cell"
      >
        Test
      </th>
    </tr>
  </thead>
</table>
`;
